kube/src/client/mod.rs

```diff
@@ -60,20 +65,30 @@ impl Client {
     /// Create and initialize a [`Client`] using the given `Service`.
     ///
     /// Use [`Client::try_from`](Self::try_from) to create with a [`Config`].
-    pub fn new<S>(service: S) -> Self
+    pub fn new<S, B>(service: S) -> Self
     where
-        S: Service<Request<Body>, Response = Response<Body>, Error = BoxError> + Send + 'static,
+        S: Service<Request<Body>, Response = Response<B>> + Send + 'static,
         S::Future: Send + 'static,
+        S::Error: Into<BoxError>,
+        B: http_body::Body<Data = bytes::Bytes> + Send + 'static,
+        B::Error: std::error::Error + Send + Sync + 'static,
     {
         Self::new_with_default_ns(service, "default")
     }
 
     /// Create and initialize a [`Client`] using the given `Service` and the default namespace.
-    fn new_with_default_ns<S, T: Into<String>>(service: S, default_ns: T) -> Self
+    fn new_with_default_ns<S, B, T: Into<String>>(service: S, default_ns: T) -> Self
     where
-        S: Service<Request<Body>, Response = Response<Body>, Error = BoxError> + Send + 'static,
+        S: Service<Request<Body>, Response = Response<B>> + Send + 'static,
         S::Future: Send + 'static,
+        S::Error: Into<BoxError>,
+        B: http_body::Body<Data = bytes::Bytes> + Send + 'static,
+        B::Error: std::error::Error + Send + Sync + 'static,
     {
+        // Transform response body to `hyper::Body` and use type erased error to avoid type parameters.
+        let service = MapResponseBodyLayer::new(|b| hyper::Body::wrap_stream(body::IntoStream::new(b)))
+            .layer(service)
+            .map_err(|e| e.into());
```

Comment:
Took me a few days to come up with this, but should be worth it. This change allows us to take advantage of the Tower ecosystem much more without introducing type parameters.

Shouldn't be a breaking change, but need to test more to make sure.